server/core/redisCache.js

```
@@ -0,0 +1,75 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+/**
+ * Creates the redis caching manager and client and provides the helper function `cacheInRedis`.
+ *
+ * The function `cacheInRedis`  wraps fetches and caches the values returned.
+ * The `wrap` helper from cache-manager first checks the cache to see if the metrics with the given key
+ * are already in cache and not. If not, it calls the fetch function invokes the callback only once
+ * all files have been retrieved and cached.
+ *
+ * The callback should be a function with a signature of `function (error, results)`.
+ *
+ */
+const cacheManager = require("cache-manager");
+const redisStore = require("cache-manager-ioredis");
+const Redis = require("ioredis");
+
+const REDISHOST = process.env.REDISHOST || "localhost";
+const REDISPORT = process.env.REDISPORT || 6379;
+const REDISAUTH = process.env.REDISAUTH || "";
+
+// Expire items in cache after 1 day
+const REDIS_CACHE_TTL_SECONDS = 60 * 60 * 24;
+// Set refresh threshold to 1 hour
+const REDIS_CACHE_REFRESH_THRESHOLD = 60 * 60;
+
+const redisInstance = new Redis({
+  host: REDISHOST,
+  port: REDISPORT,
+  password: REDISAUTH,
+  ttl: REDIS_CACHE_TTL_SECONDS,
+  db: 0,
+});
+
+const redisCache = cacheManager.caching({
+  store: redisStore,
+  refreshThreshold: REDIS_CACHE_REFRESH_THRESHOLD,
+  redisInstance,
+});
+
+const redisClient = redisCache.store.getClient();
+redisClient.on("error", (error) => console.error("ERR:REDIS:", error));
```

Do we need to raise an error here in addition to logging it? What types of things constitute an error?

---

I've been thinking about this, I'm not sure if we should or not. Here's some things that constitute errors:

- TCP error (ECONNREFUSED) - The redis server is down, or some other reason a TCP connection was unsuccessful
- Redis ReplyError - Incorrect redis command, AUTH password is incorrect

In the first case for a TCP error, I don't know if we would want to raise an error, or have it default to use a memory cache if redis is not available?

---

I think we should raise an error, at least for the short term in staging just to see what errors we are receiving.

I found that you can opt in to being notified via email for the project here: https://console.cloud.google.com/user-preferences/communication

---

If it gets to be too verbose we can always change it.

---

After discussion with @daschi today, we have decided only to log the error and not to raise it at this point. I have created this issue #683  to notify Sentry directly from the BE when errors occur.